File: remarboy/src/se/remar/remarboy/LCD.java
package se.remar.remarboy;

public class LCD {

    private Memory mem;
    private int cycles = 0;

    private static final int IF = 0xFF0F;
    private static final int LCDC = 0xFF40;
    private static final int LY = 0xFF44;

    private int ly = 0;

    public LCD(Memory mem) {
        this.mem = mem;

        ly = mem.getByte(LY);
    }

    public void step() {
        cycles += mem.cycles;

        while(cycles > 114) {
            cycles -= 114;
            ly++;
            if(ly == 144) {
                mem.putByte(IF, mem.getByte(IF) | 0x01); // Vblank Interrupt Request
            }
            if(ly >= 154) {
                ly = 0;
            }
            if(ly == 0) {
                mem.putByte(IF, mem.getByte(IF) & ~0x01); // Turn off Vblank Interrupt Request
            }
            mem.putByte(LY, ly);
        }
    }

    @Override
    public String toString() {
        return "LCD, LY: "+ly + " (0x" + Util.formatByte(ly) + ")";
    }
}

File: src/core/LCD.cpp
#include "LCD.h"

const uint16_t IF = 0xFF0F;
const uint16_t LCDC = 0xFF40;
const uint16_t LY = 0xFF44;

LCD::LCD(Memory *memory) : mem(memory), cycles(0) {
  ly = mem->getByte(LY);
}

void
LCD::step() {
  cycles += mem->cycles;

  while(cycles > 114) {
    cycles -= 114;
    ly++;
    if(ly == 144) {
      // Vblank Interrupt Request
      mem->putByte(IF, mem->getByte(IF) | 0x01);
    }
    if(ly >= 154) {
      ly = 0;
    }
    if(ly == 0) {
      // Turn off Vblank Interrupt Request
      mem->putByte(IF, mem->getByte(IF) & ~0x01);
    }
    mem->putByte(LY, ly);
  }
}
